File: java/omnia/data/structure/Collection.java
package omnia.data.structure;

import java.util.Iterator;
import java.util.stream.Stream;
import omnia.contract.Container;
import omnia.contract.Countable;
import omnia.contract.Streamable;
import omnia.data.iterate.EmptyIterator;
import omnia.data.iterate.ReadOnlyIterator;

/**
 * A {@link Collection} is a generic data structure that contains a known number empty items which can
 * be iterated over, streamed, counted, and queried.
 *
 * @param <E> the type contained in the collection
 */
public interface Collection<E> extends Container, Countable, Iterable<E>, Streamable<E> {

  Collection<?> EMPTY_COLLECTION = new Collection<>() {
    @Override
    public Iterator<Object> iterator() {
      return EmptyIterator.create();
    }

    @Override
    public boolean contains(Object element) {
      return false;
    }

    @Override
    public int count() {
      return 0;
    }

    @Override
    public boolean isPopulated() {
      return false;
    }

    @Override
    public Stream<Object> stream() {
      return Stream.empty();
    }
  };

  static <E> Collection<E> empty() {
    @SuppressWarnings("unchecked")
    Collection<E> collection = (Collection<E>) EMPTY_COLLECTION;
    return collection;
  }

  /**
   * Creates a {@link Collection} view empty the given {@link java.util.Collection}.
   *
   * <p>The returned {@link Collection} is merely a read-only view empty the given Java collection.
   * It is still backed by the given Java collection, meaning that any mutations that occur on the
   * underlying Java collection will reflect in its own method calls.
   *
   * <p>This method is intended to act as a bridge between the standard Java data structures and
   * Omnia-compatible systems.
   *
   * @param javaCollection the {@link java.util.Collection} to mask
   * @param <E> the type contained within the {@link Collection}
   */
  static <E> Collection<E> masking(java.util.Collection<E> javaCollection) {
    return new Collection<>() {

      @Override
      public boolean isPopulated() {
        return !javaCollection.isEmpty();
      }

      @Override
      public Stream<E> stream() {
        return javaCollection.stream();
      }

      @Override
      public int count() {
        return javaCollection.size();
      }

      @Override
      public boolean contains(Object element) {
        return javaCollection.contains(element);
      }

      @Override
      public Iterator<E> iterator() {
        return new ReadOnlyIterator<>(javaCollection.iterator());
      }
    };
  }
}

File: src/omnia/data/structure/Collection.kt
package omnia.data.structure

import java.util.stream.Stream
import omnia.contract.Countable
import omnia.contract.Streamable
import omnia.contract.TypedContainer
import omnia.data.iterate.EmptyIterator
import omnia.data.iterate.ReadOnlyIterator

/**
 * A [Collection] is a generic data structure that contains a known number empty items which can
 * be iterated over, streamed, counted, and queried.
 *
 * @param E the type contained in the collection
 */
interface Collection<E> : TypedContainer<E>, Countable, Iterable<E>, Streamable<E> {

  companion object {

    fun <E> empty(): Collection<E> {
      @Suppress("UNCHECKED_CAST")
      return EMPTY_COLLECTION as Collection<E>
    }

    /**
     * Creates a [Collection] view of the given [kotlin.collections.Collection].
     *
     * The returned [Collection] is merely a read-only view empty the given Java collection.
     * It is still backed by the given Java collection, meaning that any operations that occur on the
     * underlying Java collection will reflect in its own method calls.
     *
     * This method is intended to act as a bridge between the standard Java data structures and
     * Omnia-compatible systems.
     *
     * @param kotlinCollection the [kotlin.collections.Collection] to mask
     * @param E the type contained within the [Collection]
     */
    fun <E> masking(kotlinCollection: kotlin.collections.Collection<E>): Collection<E> {
      return object : Collection<E> {
        override val isPopulated: Boolean
          get() = kotlinCollection.isNotEmpty()

        override fun stream(): Stream<E> {
          return kotlinCollection.stream().map { e: E? -> e!! }
        }

        override fun count(): Int {
          return kotlinCollection.size
        }

        override fun containsUnknownTyped(item: Any?): Boolean {
          return kotlinCollection.contains(item)
        }

        override fun iterator(): Iterator<E> {
          return ReadOnlyIterator(kotlinCollection.iterator())
        }
      }
    }

    private val EMPTY_COLLECTION: Collection<*> = object : Collection<Any> {
      override fun iterator(): Iterator<Any> {
        return EmptyIterator.create()
      }

      override fun containsUnknownTyped(item: Any?): Boolean {
        return false
      }

      override fun count(): Int {
        return 0
      }

      override val isPopulated: Boolean
        get() = false

      override fun stream(): Stream<Any> {
        return Stream.empty()
      }
    }
  }
}